<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>SARUMAN: /home/tobias/work/SunStudioProjects/saruman/src/util.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css">
<link href="doxygen.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.8 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul>
      <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
      <li><a href="annotated.html"><span>Data&nbsp;Structures</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
    </ul>
  </div>
  <div class="tabs">
    <ul>
      <li><a href="files.html"><span>File&nbsp;List</span></a></li>
      <li><a href="globals.html"><span>Globals</span></a></li>
    </ul>
  </div>
</div>
<div class="contents">
<h1>/home/tobias/work/SunStudioProjects/saruman/src/util.c File Reference</h1><code>#include &lt;stdio.h&gt;</code><br>
<code>#include &lt;ctype.h&gt;</code><br>
<code>#include &lt;string.h&gt;</code><br>
<code>#include &lt;stdlib.h&gt;</code><br>
<code>#include &quot;argtable2.h&quot;</code><br>
<code>#include &quot;<a class="el" href="util_8h-source.html">util.h</a>&quot;</code><br>
<code>#include &quot;<a class="el" href="saruman_8h-source.html">saruman.h</a>&quot;</code><br>
<code>#include &quot;<a class="el" href="hash_8h-source.html">hash.h</a>&quot;</code><br>
<table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Defines</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">#define&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#1c6d5de492ac61ad29aec7aa9a436bbf">VERSION</a>&nbsp;&nbsp;&nbsp;&quot;SVN-BUILD-M2&quot;</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">#define&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#095f50fa15c6a279dd92d579f21e1cd1">SWAP_CHAR</a>(x, y)&nbsp;&nbsp;&nbsp;{char c; c = x; x = y; y = c;}</td></tr>

<tr><td colspan="2"><br><h2>Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#cf3c7abd00b7ed252afccc0a4474a80f">reverse</a> (char *sequence, int sequence_length)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#ee13853bdebdbeef1f20ec7dec955baa">reverse_complement</a> (char *sequence, int sequence_length)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">signed short int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#99fd0b758cd04342a6887e657d634de7">check_user_input</a> (int argc, char *argv[])</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#91efbe26ec264f7701b4e4cab34c2331">read_multiple_fasta_file</a> (const char *filename, int max_entries, int readlength)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="util_8c.html#f7184965942e6545edb4af31cc12688d">construct_qgram_index</a> (char *<a class="el" href="saruman_8h.html#571c0bdece58dc19b6806fae78c2cfa6">genome_sequence</a>, int qgram_size, int run)</td></tr>

</table>
<hr><h2>Define Documentation</h2>
<a class="anchor" name="095f50fa15c6a279dd92d579f21e1cd1"></a><!-- doxytag: member="util.c::SWAP_CHAR" ref="095f50fa15c6a279dd92d579f21e1cd1" args="(x, y)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define SWAP_CHAR          </td>
          <td>(</td>
          <td class="paramtype">x,         <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">y&nbsp;</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td>&nbsp;&nbsp;&nbsp;{char c; c = x; x = y; y = c;}</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

</div>
</div><p>
<a class="anchor" name="1c6d5de492ac61ad29aec7aa9a436bbf"></a><!-- doxytag: member="util.c::VERSION" ref="1c6d5de492ac61ad29aec7aa9a436bbf" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VERSION&nbsp;&nbsp;&nbsp;&quot;SVN-BUILD-M2&quot;          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

<p>Referenced by <a class="el" href="util_8c-source.html#l00060">check_user_input()</a>.</p>

</div>
</div><p>
<hr><h2>Function Documentation</h2>
<a class="anchor" name="99fd0b758cd04342a6887e657d634de7"></a><!-- doxytag: member="util.c::check_user_input" ref="99fd0b758cd04342a6887e657d634de7" args="(int argc, char *argv[])" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">signed short int check_user_input           </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>argc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>argv</em>[]</td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Checks the user input. Tests if all mandatory values are given. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>argc</em>&nbsp;</td><td>number of CLI arguments </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>argv</em>&nbsp;</td><td>char array of the CLI arguments </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>0 program starrted but directly terminated (show help, version etc)<br>
 -1 missing mandaroty value<br>
 2 all mandatory values ok, start mapping programm </dd></dl>

<p>References <a class="el" href="util_8h-source.html#l00042">be_verbose_parameter</a>, <a class="el" href="util_8h-source.html#l00025">be_verbose_value</a>, <a class="el" href="util_8h-source.html#l00034">chunk_parameter</a>, <a class="el" href="util_8h-source.html#l00019">chunk_value</a>, <a class="el" href="util_8h-source.html#l00047">end</a>, <a class="el" href="util_8h-source.html#l00033">error_rate_parameter</a>, <a class="el" href="util_8h-source.html#l00018">error_rate_value</a>, <a class="el" href="util_8h-source.html#l00045">fastq_parameter</a>, <a class="el" href="util_8h-source.html#l00041">genome_chunk_parameter</a>, <a class="el" href="util_8h-source.html#l00023">genome_chunk_value</a>, <a class="el" href="util_8h-source.html#l00039">genome_file_parameter</a>, <a class="el" href="util_8h-source.html#l00016">genome_file_value</a>, <a class="el" href="util_8h-source.html#l00043">help_parameter</a>, <a class="el" href="util_8h-source.html#l00027">print_non_matching_reads</a>, <a class="el" href="util_8h-source.html#l00046">print_non_matching_reads_parameter</a>, <a class="el" href="util_8h-source.html#l00040">read_file_parameter</a>, <a class="el" href="util_8h-source.html#l00017">read_file_value</a>, <a class="el" href="util_8h-source.html#l00038">read_length_parameter</a>, <a class="el" href="util_8h-source.html#l00024">read_length_value</a>, <a class="el" href="util_8h-source.html#l00037">read_number_parameter</a>, <a class="el" href="util_8h-source.html#l00022">read_number_value</a>, <a class="el" href="util_8h-source.html#l00035">sw_match_parameter</a>, <a class="el" href="util_8h-source.html#l00020">sw_match_value</a>, <a class="el" href="util_8h-source.html#l00036">sw_mismatch_parameter</a>, <a class="el" href="util_8h-source.html#l00021">sw_mismatch_value</a>, <a class="el" href="util_8h-source.html#l00026">use_fastq</a>, <a class="el" href="util_8c-source.html#l00019">VERSION</a>, and <a class="el" href="util_8h-source.html#l00044">version_parameter</a>.</p>

<p>Referenced by <a class="el" href="saruman_8c-source.html#l00028">main()</a>.</p>

</div>
</div><p>
<a class="anchor" name="f7184965942e6545edb4af31cc12688d"></a><!-- doxytag: member="util.c::construct_qgram_index" ref="f7184965942e6545edb4af31cc12688d" args="(char *genome_sequence, int qgram_size, int run)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void construct_qgram_index           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>genome_sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>qgram_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>run</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructs the main qgram index (<a class="el" href="structqgram__hash.html" title="a structure needed to built a hash index of qgram start positions">qgram_hash</a>) holding all start positions for each qgram in the genome <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>genome_sequence</em>&nbsp;</td><td>the genome sequence returned by <a class="el" href="util_8h.html#56acbbc313fd2b9f268831be66e30ed5">read_genome_fasta_file</a> </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>qgram_size</em>&nbsp;</td><td>the desired qgram size (computed automaticly as <code>short int qgram_length = ceil(read_length / (maximal_errors_allowed+1))-1</code>) </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>nothing </dd></dl>

<p>References <a class="el" href="hash_8c-source.html#l00167">add_qgram_to_index()</a>, and <a class="el" href="util_8h-source.html#l00023">genome_chunk_value</a>.</p>

<p>Referenced by <a class="el" href="saruman_8c-source.html#l00028">main()</a>.</p>

</div>
</div><p>
<a class="anchor" name="91efbe26ec264f7701b4e4cab34c2331"></a><!-- doxytag: member="util.c::read_multiple_fasta_file" ref="91efbe26ec264f7701b4e4cab34c2331" args="(const char *filename, int max_entries, int readlength)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* read_multiple_fasta_file           </td>
          <td>(</td>
          <td class="paramtype">const char *&nbsp;</td>
          <td class="paramname"> <em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>max_entries</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>readlength</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Reads and parses the given read file. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>filename</em>&nbsp;</td><td>path to the file to read </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>max_entries</em>&nbsp;</td><td>number of reads to parse </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>readlength</em>&nbsp;</td><td>length of the reads </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>a char array containing all reads in order </dd></dl>

<p>
&lt; holds the results aka start positions in genome 
<p>References <a class="el" href="util_8h-source.html#l00031">line_offset</a>, <a class="el" href="util_8h-source.html#l00011">MAX_READ_DESCRIPTION_LENGTH</a>, <a class="el" href="saruman_8h-source.html#l00048">maximal_errors_allowed</a>, <a class="el" href="util_8h-source.html#l00060">read_name_hash::name</a>, <a class="el" href="util_8h-source.html#l00051">read_names</a>, <a class="el" href="util_8h-source.html#l00059">read_name_hash::read_number</a>, and <a class="el" href="util_8h-source.html#l00026">use_fastq</a>.</p>

<p>Referenced by <a class="el" href="saruman_8c-source.html#l00028">main()</a>.</p>

</div>
</div><p>
<a class="anchor" name="cf3c7abd00b7ed252afccc0a4474a80f"></a><!-- doxytag: member="util.c::reverse" ref="cf3c7abd00b7ed252afccc0a4474a80f" args="(char *sequence, int sequence_length)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void reverse           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>sequence_length</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Reverses a given string <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>sequence</em>&nbsp;</td><td>char sequence </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>sequence_length</em>&nbsp;</td><td>length of the sequence to be reversed </td></tr>
  </table>
</dl>
<dl class="return" compact><dt><b>Returns:</b></dt><dd>a char array containing the genome </dd></dl>

<p>References <a class="el" href="cuda-functions_8cu-source.html#l00033">SWAP_CHAR</a>.</p>

<p>Referenced by <a class="el" href="util_8c-source.html#l00036">reverse_complement()</a>.</p>

</div>
</div><p>
<a class="anchor" name="ee13853bdebdbeef1f20ec7dec955baa"></a><!-- doxytag: member="util.c::reverse_complement" ref="ee13853bdebdbeef1f20ec7dec955baa" args="(char *sequence, int sequence_length)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void reverse_complement           </td>
          <td>(</td>
          <td class="paramtype">char *&nbsp;</td>
          <td class="paramname"> <em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>sequence_length</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>

<p>References <a class="el" href="util_8c-source.html#l00023">reverse()</a>.</p>

<p>Referenced by <a class="el" href="hash_8c-source.html#l00806">prepare_cuda_alignments()</a>, <a class="el" href="hash_8c-source.html#l00280">qgram_to_read_mapping()</a>, and <a class="el" href="hash_8c-source.html#l00415">run_seashore()</a>.</p>

</div>
</div><p>
</div>
<hr size="1"><address style="text-align: right;"><small>Generated on Thu May 13 10:17:21 2010 for SARUMAN by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.8 </small></address>
</body>
</html>
